etcdserver/api/membership/cluster.go

```diff
@@ -796,44 +809,61 @@ func ValidateClusterAndAssignIDs(lg *zap.Logger, local *RaftCluster, existing *R
 	return nil
 }
 
-func mustDetectDowngrade(lg *zap.Logger, cv *semver.Version) {
+func mustDetectDowngrade(lg *zap.Logger, cv *semver.Version, d *Downgrade) {
 	lv := semver.Must(semver.NewVersion(version.Version))
 	// only keep major.minor version for comparison against cluster version
 	lv = &semver.Version{Major: lv.Major, Minor: lv.Minor}
-	if cv != nil && lv.LessThan(*cv) {
-		if IsVersionChangable(cv, lv) {
+
+	if d.Enabled && !d.TargetVersion.Equal(*lv) {
+		if lg != nil {
+			lg.Fatal(
+				"invalid downgrade; server version does not match the downgrade target version",
+				zap.String("current-server-version", version.Version),
+				zap.String("target-cluster-version", d.TargetVersion.String()),
+			)
+		} else {
+			plog.Fatalf("invalid downgrade; server version does not match the downgrade target version "+
+				"(current version: %s, target cluster version: %s).", version.Version, d.TargetVersion.String())
+		}
+		return
+	}
+
+	if cv != nil {
+		if d.Enabled && d.TargetVersion.Equal(*lv) {
 			if lg != nil {
 				lg.Info(
-					"cluster is downgrading to current server version",
-					zap.String("current-server-version", version.Version),
+					"cluster is downgrading to target version",
+					zap.String("target-cluster-version", d.TargetVersion.String()),
 					zap.String("determined-cluster-version", version.Cluster(cv.String())),
+					zap.String("current-server-version", version.Version),
 				)
 			} else {
-				plog.Infof("cluster is downgrading to current version: %s from determined cluster version: %s).", version.Version, version.Cluster(cv.String()))
+				plog.Infof("cluster is downgrading to target version: %s, "+
+					"determined cluster version: %s, current-server-version: %s).",
+					d.TargetVersion.String(), version.Cluster(cv.String()), version.Version)
 			}
-		} else {
+			return
+		}
+
+		if !d.Enabled && lv.LessThan(*cv) {
 			if lg != nil {
 				lg.Fatal(
 					"invalid downgrade; server version is lower than determined cluster version",
 					zap.String("current-server-version", version.Version),
 					zap.String("determined-cluster-version", version.Cluster(cv.String())),
 				)
 			} else {
-				plog.Fatalf("cluster cannot be downgraded (current version: %s is lower than determined cluster version: %s).", version.Version, version.Cluster(cv.String()))
+				plog.Fatalf("invalid downgrade; server version is lower than determined cluster version (current version: %s, determined cluster version: %s).",
+					version.Version, version.Cluster(cv.String()))
 			}
 		}
 	}
 }
 
 func IsVersionChangable(cv *semver.Version, lv *semver.Version) bool {
```

Comment:
Sorry I probably missed the comments when I added this in, could you help to add?
// IsVersionChangable checks the two scenario when version is changable:
// 1. Downgrade: cluster version is 1 minor version higher than local version, cluster version should change
// 2. Cluster start: when not all members version are available,  cluster version is set to MinVersion (3.0), when
//    all members are at higher version (3.2), cluster version is lower than local version, cluster version should change

Reply:
Added